File: src/__tests__/signing-vectors-manifest.test.ts
import fs from "node:fs";
import path from "node:path";
import { describe, expect, it } from "vitest";

////////////////////////////////
//  Signing-vector corpus inventory
//
//  _signing-vectors-manifest.json is the single inventory of the cross-runtime
//  signing-vector corpora (each a directory holding a vectors.json read by both a
//  vitest suite and a Kotlin twin). This suite asserts the manifest equals the set
//  of tests/fixtures/*/vectors.json directories on disk, so a new op's fixture
//  directory cannot be added without registering it here. SigningVectorsManifestTest.kt
//  reads the SAME manifest and forces every listed directory through Kotlin, so a
//  registered corpus can never be read by TS alone (ci.yml does not run the Android
//  unit tests, which is exactly the gap this closes). The protocol corpus is governed
//  by its own tests/fixtures/protocol/_manifest.json and has no vectors.json, so it is
//  outside this inventory by construction.

const FIXTURES = path.join(__dirname, "../../tests/fixtures");

const manifest = JSON.parse(fs.readFileSync(path.join(FIXTURES, "_signing-vectors-manifest.json"), "utf8")) as {
	directories: string[];
};

function dirsWithVectorsOnDisk(): string[] {
	return fs
		.readdirSync(FIXTURES, { withFileTypes: true })
		.filter((entry) => entry.isDirectory() && fs.existsSync(path.join(FIXTURES, entry.name, "vectors.json")))
		.map((entry) => entry.name)
		.sort();
}

describe("signing-vector manifest", () => {
	it("lists exactly the vectors.json directories on disk", () => {
		const listed = [...manifest.directories].sort();
		expect(listed).toEqual(dirsWithVectorsOnDisk());
	});

	it("every listed directory holds a parseable vectors.json", () => {
		for (const dir of manifest.directories) {
			const file = path.join(FIXTURES, dir, "vectors.json");
			expect(fs.existsSync(file), `${dir}/vectors.json must exist`).toBe(true);
			expect(() => JSON.parse(fs.readFileSync(file, "utf8")), `${dir}/vectors.json must parse`).not.toThrow();
		}
	});
});

File: android/app/src/test/java/com/atelier_nyaarium/switchboard/SigningVectorsManifestTest.kt
package com.atelier_nyaarium.switchboard

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Reads tests/fixtures/_signing-vectors-manifest.json (the same inventory the vitest
 * signing-vectors-manifest.test.ts reads) and forces every listed signing-vector
 * directory through Kotlin: each vectors.json must exist on the classpath and parse.
 * So a corpus registered in the manifest can never be read by TS alone - ci.yml does
 * not run these Android unit tests, so without this gate a future op's TS-only vector
 * would pass CI yet silently lack a Kotlin reader. The per-op suites (ProvisionOpsTest,
 * XDomainLinkTest, SasCryptoTest, SessionIdVectorsTest, SyncCursorVectorsTest) still own
 * the byte-level assertions; existence + parse is enough for the coverage guarantee here.
 */
class SigningVectorsManifestTest {
	private val json = Json { ignoreUnknownKeys = true }

	private fun resource(name: String): String =
		javaClass.classLoader!!.getResourceAsStream(name)!!.bufferedReader().readText()

	@Test
	fun everyListedSigningVectorCorpusParsesOnAndroid() {
		val directories = json.parseToJsonElement(resource("_signing-vectors-manifest.json"))
			.jsonObject["directories"]!!.jsonArray
		assertTrue("manifest must list at least one directory", directories.isNotEmpty())
		for (entry in directories) {
			val dir = entry.jsonPrimitive.content
			val body = resource("$dir/vectors.json")
			// Parsing throws on malformed JSON, failing the gate for that corpus.
			val parsed = json.parseToJsonElement(body).jsonObject
			assertTrue("$dir/vectors.json must be a non-empty object", parsed.isNotEmpty())
		}
	}
}
